skimage/segmentation/slic_superpixels.py

```diff
@@ -258,15 +265,15 @@ def slic(image, n_segments=100, compactness=10., max_iter=10, sigma=0,
         spacing = np.ones(3, dtype=dtype)
     elif isinstance(spacing, (list, tuple)):
         spacing = np.ascontiguousarray(spacing, dtype=dtype)
-
     if not isinstance(sigma, coll.Iterable):
         sigma = np.array([sigma, sigma, sigma], dtype=dtype)
-        sigma /= spacing.astype(dtype)
+        sigma /= spacing.astype(dtype)  # type: ignore
```

Comment:
Why this ignore?

Reply:
I could not get this to pass without the ignore. Basically, `spacing` has initially a sequence type and so this line would fail because a sequence has not `astype` method. I tried to use the `cast` function to cast `spacing` to `np.ndarray` but it did not work (much to my confusion).

Reply:
So, actually I think this is a case of mypy uncovering a potential bug in our code. If spacing is a sequence other than a tuple or list, we will get to here and have an error, because we only convert spacing (line 266 above) if it's a list or a tuple. So I in fact would suggest changing that `elif` to an `else`, and then maybe this will pass? :crossed_fingers:

Reply:
I tried but it does not work. Specifically,
```
    if spacing is None:
        spacing = np.ones(3, dtype=dtype)
    else:
        spacing = np.ascontiguousarray(spacing, dtype=dtype)

    spacing = cast(np.ndarray, spacing)
    reveal_type(spacing)
    if not isinstance(sigma, coll.Iterable):
        sigma = np.array([sigma, sigma, sigma], dtype=dtype)
        sigma /= spacing.astype(dtype)
```
and when running mypy:
```
$ mypy --ignore-missing-imports skimage/segmentation/
skimage/segmentation/slic_superpixels.py:270: note: Revealed type is 'Union[typing.Sequence[Any], None]'
skimage/segmentation/slic_superpixels.py:273: error: Item "Sequence[Any]" of "Optional[Sequence[Any]]" has no attribute "astype"
skimage/segmentation/slic_superpixels.py:273: error: Item "None" of "Optional[Sequence[Any]]" has no attribute "astype"
Found 2 errors in 1 file (checked 26 source files)
```
It's as if the `cast` was completely ignored... Any expert of mypy around here? :-p @kne42 @chrisconlan

Reply:
Thanks @emmanuelle! Well, the change should go in anyway. =)

Reply:
done

Reply:
I think the issue might be that the `spacing` variable is never declared as `np.ndarray`. The error that mypy seems to ignore for some reason is that `spacing = cast(np.ndarray, spacing)` is already a type error because the right hand side is an expression of type np.ndarray but the left hand side is a variable of type `Optional[Sequence]`.

I think the proper solution is to split the spacing variable into two, one with type  `Optional[Sequence]` the other of type `np.ndarray`.